dom5.ts

```diff
@@ -605,6 +605,37 @@ export function remove(node: Node) {
   node.parentNode = undefined;
 }
 
+/**
+ * Removes a node and places its children in its place.  If the node
+ * has no parent, the operation is impossible and no action takes place.
+ */
+export function removeNodeSaveChildren(node: Node) {
+  // We can't save the children if there's no parent node to provide
+  // for them.
+  const fosterParent = node.parentNode;
+  if (!fosterParent) {
+    return;
+  }
+  const children = (node.childNodes || []).slice();
+  for (const child of children) {
+    insertBefore(node.parentNode!, node, child);
+  }
+  remove(node);
+}
+
+/**
+ * When parse5 parses an HTML document with `parse`, it injects missing root
+ * elements (html, head and body) if they are missing.  This function removes
+ * these from the AST if they have no location info, so it requires that
+ * the `parse5.parse` be used with the `locationInfo` option of `true`.
+ */
+export function removeFakeRootElements(ast: Node) {
+  const injectedNodes = queryAll(
+      ast,
+      (n) => !n.__location && hasMatchingTagName(/^(html|head|body)$/i)(n));
+  injectedNodes.reverse().forEach(removeNodeSaveChildren);
```

Comment:
We could avoid walking the entire tree at this point, only looking 2-3 levels deep I think